### Class 7- Starter code


In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model, metrics


### Create sample data and fit a model

In [2]:
df = pd.DataFrame({'x': range(100), 'y': range(100)})
biased_df  = df.copy()
biased_df.loc[:20, 'x'] = 1
biased_df.loc[:20, 'y'] = 1

def append_jitter(series):
    jitter = np.random.random_sample(size=100)
    return series + jitter

df['x'] = append_jitter(df.x)
df['y'] = append_jitter(df.y)
biased_df['x'] = append_jitter(biased_df.x)
biased_df['y'] = append_jitter(biased_df.y)

In [3]:
biased_df.head(100)

x          y
0    1.750282   1.951180
1    1.506911   1.045775
2    1.169209   1.170346
3    1.886087   1.851784
4    1.561639   1.543202
5    1.533204   1.664606
6    1.168309   1.720108
7    1.179142   1.548226
8    1.214172   1.335108
9    1.037930   1.130273
10   1.833000   1.499473
11   1.275291   1.484812
12   1.353787   1.012120
13   1.183090   1.676030
14   1.420423   1.584590
15   1.108716   1.764610
16   1.856325   1.615629
17   1.569148   1.738790
18   1.890486   1.436285
19   1.249997   1.974582
20   1.280985   1.123163
21  21.733087  21.481995
22  22.219351  22.382321
23  23.324507  23.839001
24  24.764464  24.369157
25  25.355407  25.845794
26  26.991617  26.547688
27  27.705229  27.233806
28  28.462196  28.842593
29  29.638374  29.404341
..        ...        ...
70  70.215732  70.732138
71  71.428889  71.342185
72  72.838354  72.107429
73  73.984257  73.359161
74  74.955419  74.630277
75  75.292494  75.458258
76  76.786353  76.398929
77  77.449865  77.355291
78  78.555451  78.914037
79  79.859355  79.327326
80  80.988684  80.663809
81  81.080732  81.889685
82  82.712939  82.716858
83  83.736257  83.490387
84  84.509252  84.494154
85  85.328375  85.750700
86  86.335492  86.544687
87  87.483904  87.864452
88  88.361451  88.651709
89  89.350325  89.027892
90  90.483577  90.312842
91  91.865670  91.768777
92  92.597733  92.418247
93  93.039364  93.058434
94  94.122996  94.740048
95  95.303828  95.096358
96  96.615856  96.498413
97  97.656019  97.282845
98  98.359126  98.971066
99  99.191198  99.552926

[100 rows x 2 columns]

In [4]:
np.abs(((df['y']-df['x'])**2).sum())

16.223192569742935

In [5]:
## fit
lm = linear_model.LinearRegression().fit(df[['x']], df['y'])
print metrics.mean_squared_error(df['y'], lm.predict(df[['x']]))

0.156618552272


In [6]:
## biased fit
lm = linear_model.LinearRegression().fit(biased_df[['x']], biased_df['y'])
print metrics.mean_squared_error(biased_df['y'], lm.predict(biased_df[['x']]))

0.151233333886


## Cross validation
#### Intro to cross validation with bike share data from last time. We will be modeling casual ridership. 

In [7]:
from sklearn import cross_validation
wd = '../../assets/dataset/'
bikeshare = pd.read_csv(wd + 'bikeshare.csv')

C:\Users\Andrew\Anaconda2\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


#### Create dummy variables and set outcome (dependent) variable

In [58]:
weather = pd.get_dummies(bikeshare.weathersit, prefix='weather')
modeldata = bikeshare[['temp', 'hum']].join(weather[['weather_1', 'weather_2', 'weather_3']])
y = bikeshare.casual 

In [59]:
modeldata

temp   hum  weather_1  weather_2  weather_3
0      0.24  0.81          1          0          0
1      0.22  0.80          1          0          0
2      0.22  0.80          1          0          0
3      0.24  0.75          1          0          0
4      0.24  0.75          1          0          0
5      0.24  0.75          0          1          0
6      0.22  0.80          1          0          0
7      0.20  0.86          1          0          0
8      0.24  0.75          1          0          0
9      0.32  0.76          1          0          0
10     0.38  0.76          1          0          0
11     0.36  0.81          1          0          0
12     0.42  0.77          1          0          0
13     0.46  0.72          0          1          0
14     0.46  0.72          0          1          0
15     0.44  0.77          0          1          0
16     0.42  0.82          0          1          0
17     0.44  0.82          0          1          0
18     0.42  0.88          0          0          1
19     0.42  0.88          0          0          1
20     0.40  0.87          0          1          0
21     0.40  0.87          0          1          0
22     0.40  0.94          0          1          0
23     0.46  0.88          0          1          0
24     0.46  0.88          0          1          0
25     0.44  0.94          0          1          0
26     0.42  1.00          0          1          0
27     0.46  0.94          0          1          0
28     0.46  0.94          0          1          0
29     0.42  0.77          0          0          1
...     ...   ...        ...        ...        ...
17349  0.24  0.44          0          1          0
17350  0.34  0.61          1          0          0
17351  0.22  0.47          1          0          0
17352  0.20  0.51          1          0          0
17353  0.20  0.55          1          0          0
17354  0.20  0.51          1          0          0
17355  0.18  0.55          1          0          0
17356  0.18  0.55          1          0          0
17357  0.16  0.59          1          0          0
17358  0.16  0.59          1          0          0
17359  0.14  0.69          1          0          0
17360  0.16  0.64          1          0          0
17361  0.16  0.64          1          0          0
17362  0.16  0.64          1          0          0
17363  0.14  0.69          1          0          0
17364  0.18  0.64          0          1          0
17365  0.20  0.69          0          1          0
17366  0.22  0.60          0          1          0
17367  0.24  0.56          0          1          0
17368  0.26  0.44          0          1          0
17369  0.28  0.45          0          1          0
17370  0.28  0.45          0          1          0
17371  0.26  0.48          0          1          0
17372  0.26  0.48          0          1          0
17373  0.26  0.48          0          1          0
17374  0.26  0.60          0          1          0
17375  0.26  0.60          0          1          0
17376  0.26  0.60          1          0          0
17377  0.26  0.56          1          0          0
17378  0.26  0.65          1          0          0

[17379 rows x 5 columns]

#### Create a cross valiation with 5 folds

In [9]:
for k in range(2,10,1):
    #k=5
    kf = cross_validation.KFold(len(modeldata), n_folds=k, shuffle=True)
    mse_values_train = []
    mse_values = []
    scores = []
    n=0
    for train_index, test_index in kf:
        lm = linear_model.LinearRegression().fit(modeldata.iloc[train_index], y.iloc[train_index])
        mse_values_train.append(metrics.mean_squared_error(y.iloc[train_index], lm.predict(modeldata.iloc[train_index])))
        mse_values.append(metrics.mean_squared_error(y.iloc[test_index], lm.predict(modeldata.iloc[test_index])))
        scores.append(lm.score(modeldata, y))
        print train_index, test_index, mse_values_train[n].round(6), mse_values[n].round(6), scores[n].round(6)
        n += 1
    print "~~~~ SUMMARY OF CROSS VALIDATION {} FOLDS~~~~".format(k)
    print np.mean(mse_values_train)
    print 'Mean of MSE for all folds:', np.mean(mse_values)
    print 'Mean of R2 for all folds:', np.mean(scores)
    print

[    8     9    10 ..., 17373 17374 17376] [    0     1     2 ..., 17375 17377 17378] 1680.377167 1665.356378 0.311817
[    0     1     2 ..., 17375 17377 17378] [    8     9    10 ..., 17373 17374 17376] 1664.202097 1681.559598 0.311812
~~~~ SUMMARY OF CROSS VALIDATION 2 FOLDS~~~~
1672.28963177
Mean of MSE for all folds: 1673.45798823
Mean of R2 for all folds: 0.311814385883

[    0     1     2 ..., 17373 17376 17377] [   11    15    18 ..., 17374 17375 17378] 1676.470628 1665.379847 0.311855
[    0     1     4 ..., 17376 17377 17378] [    2     3     6 ..., 17368 17372 17373] 1698.15863 1621.702928 0.311897
[    2     3     6 ..., 17374 17375 17378] [    0     1     4 ..., 17371 17376 17377] 1642.73922 1732.535169 0.311898
~~~~ SUMMARY OF CROSS VALIDATION 3 FOLDS~~~~
1672.45615922
Mean of MSE for all folds: 1673.20598131
Mean of R2 for all folds: 0.311883186747

[    0     1     2 ..., 17375 17376 17378] [    4     9    13 ..., 17362 17366 17377] 1681.527358 1646.565447 0.31185
[    

In [10]:
kf = cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True)

In [11]:
mse_values = []
scores = []
n= 0
print "~~~~ CROSS VALIDATION each fold ~~~~"
for train_index, test_index in kf:
    lm = linear_model.LinearRegression().fit(modeldata.iloc[train_index], y.iloc[train_index])
    mse_values.append(metrics.mean_squared_error(y.iloc[test_index], lm.predict(modeldata.iloc[test_index])))
    scores.append(lm.score(modeldata, y))
    n+=1
    print 'Model', n
    print 'MSE:', mse_values[n-1]
    print 'R2:', scores[n-1]


print "~~~~ SUMMARY OF CROSS VALIDATION ~~~~"
print 'Mean of MSE for all folds:', np.mean(mse_values)
print 'Mean of R2 for all folds:', np.mean(scores)

~~~~ CROSS VALIDATION each fold ~~~~
Model 1
MSE: 1865.80650524
R2: 0.311904370208
Model 2
MSE: 1655.81315233
R2: 0.311914626525
Model 3
MSE: 1600.73968669
R2: 0.311904375989
Model 4
MSE: 1762.80043928
R2: 0.311842842944
Model 5
MSE: 1482.73868172
R2: 0.311876047373
~~~~ SUMMARY OF CROSS VALIDATION ~~~~
Mean of MSE for all folds: 1673.57969305
Mean of R2 for all folds: 0.311888452608


In [12]:
lm = linear_model.LinearRegression().fit(modeldata, y)
print "~~~~ Single Model ~~~~"
print 'MSE of single model:', metrics.mean_squared_error(y, lm.predict(modeldata))
print 'R2: ', lm.score(modeldata, y)

~~~~ Single Model ~~~~
MSE of single model: 1672.58110765
R2:  0.311934605989


### Check
While the cross validated approach here generated more overall error, which of the two approaches would predict new data more accurately: the single model or the cross validated, averaged one? Why?


Answer: Cross validated, averaged one. It is just safer in avoiding the overfitting problem.

### There are ways to improve our model with regularization. 
Let's check out the effects on MSE and R2

In [13]:
lm = linear_model.LinearRegression().fit(modeldata, y)
print "~~~ OLS ~~~"
print 'OLS MSE: ', metrics.mean_squared_error(y, lm.predict(modeldata))
print 'OLS R2:', lm.score(modeldata, y)

lm = linear_model.Lasso().fit(modeldata, y)
print "~~~ Lasso ~~~"
print 'Lasso MSE: ', metrics.mean_squared_error(y, lm.predict(modeldata))
print 'Lasso R2:', lm.score(modeldata, y)

lm = linear_model.Ridge().fit(modeldata, y)
print "~~~ Ridge ~~~"
print 'Ridge MSE: ', metrics.mean_squared_error(y, lm.predict(modeldata))
print 'Ridge R2:', lm.score(modeldata, y)

~~~ OLS ~~~
OLS MSE:  1672.58110765
OLS R2: 0.311934605989
~~~ Lasso ~~~
Lasso MSE:  1725.41581608
Lasso R2: 0.290199495922
~~~ Ridge ~~~
Ridge MSE:  1672.60490113
Ridge R2: 0.311924817843


### Figuring out the alphas can be done by "hand"

In [14]:
alphas = np.logspace(-10, 10, 21)
for a in alphas:
    print 'Alpha:', a
    lm = linear_model.Ridge(alpha=a)
    lm.fit(modeldata, y)
    print lm.coef_
    print metrics.mean_squared_error(y, lm.predict(modeldata))

Alpha: 1e-10
[ 112.68901765  -84.01121684  -24.68489063  -21.00314493  -21.71893628]
1672.58110765
Alpha: 1e-09
[ 112.68901765  -84.01121684  -24.68489061  -21.00314491  -21.71893626]
1672.58110765
Alpha: 1e-08
[ 112.68901765  -84.01121684  -24.6848904   -21.00314471  -21.71893606]
1672.58110765
Alpha: 1e-07
[ 112.68901763  -84.01121682  -24.68488837  -21.00314268  -21.71893403]
1672.58110765
Alpha: 1e-06
[ 112.68901745  -84.01121667  -24.68486804  -21.00312237  -21.71891373]
1672.58110765
Alpha: 1e-05
[ 112.68901562  -84.01121509  -24.68466472  -21.00291929  -21.71871079]
1672.58110765
Alpha: 0.0001
[ 112.68899732  -84.01119938  -24.68263174  -21.00088873  -21.71668162]
1672.58110765
Alpha: 0.001
[ 112.68881437  -84.01104228  -24.66232204  -20.98060316  -21.69640993]
1672.58110774
Alpha: 0.01
[ 112.68698753  -84.00947323  -24.46121539  -20.77973778  -21.49568404]
1672.58111645
Alpha: 0.1
[ 112.66896732  -83.99396383  -22.63109556  -18.95202277  -19.66942371]
1672.58185208
Alpha: 1.0
[

### Or we can use grid search to make this faster

In [15]:
from sklearn import grid_search

alphas = np.logspace(-10, 10, 21)
gs = grid_search.GridSearchCV(
    estimator=linear_model.Ridge(),
    param_grid={'alpha': alphas},
    scoring='neg_mean_squared_error')

gs.fit(modeldata, y)


C:\Users\Andrew\Anaconda2\lib\site-packages\sklearn\grid_search.py:43: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


GridSearchCV(cv=None, error_score='raise',
       estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'alpha': array([  1.00000e-10,   1.00000e-09,   1.00000e-08,   1.00000e-07,
         1.00000e-06,   1.00000e-05,   1.00000e-04,   1.00000e-03,
         1.00000e-02,   1.00000e-01,   1.00000e+00,   1.00000e+01,
         1.00000e+02,   1.00000e+03,   1.00000e+04,   1.00000e+05,
         1.00000e+06,   1.00000e+07,   1.00000e+08,   1.00000e+09,
         1.00000e+10])},
       pre_dispatch='2*n_jobs', refit=True,
       scoring='neg_mean_squared_error', verbose=0)

##### Best score 

In [16]:
gs.best_score_ 

-1814.0936913337962

##### mean squared error here comes in negative, so let's make it positive.

In [17]:
-gs.best_score_ 

1814.0936913337962

##### explains which grid_search setup worked best

In [18]:
gs.best_estimator_ 

Ridge(alpha=10.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

##### shows all the grid pairings and their performances.

In [19]:
for s in gs.grid_scores_:
    print s

mean: -1817.58711, std: 542.14315, params: {'alpha': 1e-10}
mean: -1817.58711, std: 542.14315, params: {'alpha': 1.0000000000000001e-09}
mean: -1817.58711, std: 542.14315, params: {'alpha': 1e-08}
mean: -1817.58711, std: 542.14315, params: {'alpha': 9.9999999999999995e-08}
mean: -1817.58711, std: 542.14315, params: {'alpha': 9.9999999999999995e-07}
mean: -1817.58711, std: 542.14317, params: {'alpha': 1.0000000000000001e-05}
mean: -1817.58707, std: 542.14331, params: {'alpha': 0.0001}
mean: -1817.58663, std: 542.14477, params: {'alpha': 0.001}
mean: -1817.58230, std: 542.15933, params: {'alpha': 0.01}
mean: -1817.54318, std: 542.30102, params: {'alpha': 0.10000000000000001}
mean: -1817.20111, std: 543.63587, params: {'alpha': 1.0}
mean: -1814.09369, std: 556.35563, params: {'alpha': 10.0}
mean: -1818.51694, std: 653.68607, params: {'alpha': 100.0}
mean: -2125.58777, std: 872.45270, params: {'alpha': 1000.0}
mean: -2458.08836, std: 951.30428, params: {'alpha': 10000.0}
mean: -2532.21151,

### ACTIVITY

## Gradient Descent

In [20]:
num_to_approach, start, steps, optimized = 6.5, 0., [-1, 1], False
while not optimized:
    current_distance = num_to_approach - start
    got_better = False
    next_steps = [start + i for i in steps]
    for n in next_steps:
        print n
        distance = np.abs(num_to_approach - n)
        if distance < current_distance:
            got_better = True
            print distance, 'is better than', current_distance
            current_distance = distance
            start = n
    if got_better:
        print 'found better solution! using', current_distance
    else:
        optimized = True
        print start, 'is closest to', num_to_approach


-1.0
1.0
5.5 is better than 6.5
found better solution! using 5.5
0.0
2.0
4.5 is better than 5.5
found better solution! using 4.5
1.0
3.0
3.5 is better than 4.5
found better solution! using 3.5
2.0
4.0
2.5 is better than 3.5
found better solution! using 2.5
3.0
5.0
1.5 is better than 2.5
found better solution! using 1.5
4.0
6.0
0.5 is better than 1.5
found better solution! using 0.5
5.0
7.0
6.0 is closest to 6.5


### Bonus: 
implement a stopping point, similar to what n_iter would do in gradient descent when we've reached "good enough"

In [21]:
num_to_approach, start, steps, optimized = 6.5, 0., [-1, 1], False
n_iter = 0
while not optimized:
    n_iter += 1
    current_distance = np.abs(num_to_approach - start)
    got_better = False
    next_steps = [start + i for i in steps]
    print n_iter
    for n in next_steps:
        #print n
        distance = np.abs(num_to_approach - n)
        if distance <= current_distance:   # deliberately making the logic wrong here
            got_better = True
            print distance, 'is better than', current_distance
            current_distance = distance
            start = n
        #print n_iter,distance,current_distance
    if (got_better) & (n_iter < 1000):     # adding extra criterion to avoid infinite loop
        print 'found better solution! using distance =', current_distance, 'and start =',start
                # elif (distance == current_distance) & (n_iter < 1000):
                #     print 'continue finding closer solution until 1000th iteration'
    else:        
        optimized = True
        print start, 'is closest to', num_to_approach


1
5.5 is better than 6.5
found better solution! using distance = 5.5 and start = 1.0
2
4.5 is better than 5.5
found better solution! using distance = 4.5 and start = 2.0
3
3.5 is better than 4.5
found better solution! using distance = 3.5 and start = 3.0
4
2.5 is better than 3.5
found better solution! using distance = 2.5 and start = 4.0
5
1.5 is better than 2.5
found better solution! using distance = 1.5 and start = 5.0
6
0.5 is better than 1.5
found better solution! using distance = 0.5 and start = 6.0
7
0.5 is better than 0.5
found better solution! using distance = 0.5 and start = 7.0
8
0.5 is better than 0.5
found better solution! using distance = 0.5 and start = 6.0
9
0.5 is better than 0.5
found better solution! using distance = 0.5 and start = 7.0
10
0.5 is better than 0.5
found better solution! using distance = 0.5 and start = 6.0
11
0.5 is better than 0.5
found better solution! using distance = 0.5 and start = 7.0
12
0.5 is better than 0.5
found better solution! using distance

## Demo: Application of Gradient Descent 

In [22]:
lm = linear_model.SGDRegressor()
lm.fit(modeldata, y)
print "Gradient Descent R2:", lm.score(modeldata, y)
print "Gradient Descent MSE:", metrics.mean_squared_error(y, lm.predict(modeldata))

Gradient Descent R2: 0.307877801382
Gradient Descent MSE: 1682.44257548


### Check: Untuned, how well did gradient descent perform compared to OLS?

Answer: With lower R-squared and higher MSE, we can say that the performance of gradient descent is not as good as OLS.

# Independent Practice: Bike data revisited

There are tons of ways to approach a regression problem. The regularization techniques appended to ordinary least squares optimizes the size of coefficients to best account for error. Gradient Descent also introduces learning rate (how aggressively do we solve the problem), epsilon (at what point do we say the error margin is acceptable), and iterations (when should we stop no matter what?)

For this deliverable, our goals are to:

- implement the gradient descent approach to our bike-share modeling problem,
- show how gradient descent solves and optimizes the solution,
- demonstrate the grid_search module!

While exploring the Gradient Descent regressor object, you'll build a grid search using the stochastic gradient descent estimator for the bike-share data set. Continue with either the model you evaluated last class or the simpler one from today. In particular, be sure to implement the "param_grid" in the grid search to get answers for the following questions:

- With a set of alpha values between 10^-10 and 10^-1, how does the mean squared error change?
- Based on the data, we know when to properly use L1 vs L2 regularization. By using a grid search with L1_ratios between 0 and 1 (increasing every 0.05), does that statement hold true? If not, did gradient descent have enough iterations?
- How do these results change when you alter the learning rate (eta0)?

**Bonus**: Can you see the advantages and disadvantages of using gradient descent after finishing this exercise?

### Starter Code

In [53]:
alphas = np.logspace(-10, -1, 10)
params = {'alpha': alphas} # put your gradient descent parameters here
print params
gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=5),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l2 in gs.grid_scores_:
    print l2

{'alpha': array([  1.00000000e-10,   1.00000000e-09,   1.00000000e-08,
         1.00000000e-07,   1.00000000e-06,   1.00000000e-05,
         1.00000000e-04,   1.00000000e-03,   1.00000000e-02,
         1.00000000e-01])}
BEST ESTIMATOR
1688.90980088
SGDRegressor(alpha=9.9999999999999995e-08, average=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='invscaling', loss='squared_loss', n_iter=5,
       penalty='l2', power_t=0.25, random_state=None, shuffle=True,
       verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -1688.95769, std: 98.23973, params: {'alpha': 1e-10}
mean: -1691.59497, std: 94.02736, params: {'alpha': 1.0000000000000001e-09}
mean: -1689.09775, std: 97.04417, params: {'alpha': 1e-08}
mean: -1688.90980, std: 97.22039, params: {'alpha': 9.9999999999999995e-08}
mean: -1689.35137, std: 97.42058, params: {'alpha': 9.9999999999999995e-07}
mean: -1689.72196, std: 96.26277, params: {'alpha': 1.0000000000000001e-05}
mean: -1691.12843, s

In [54]:
alphas = np.logspace(-10, -1, 10)
params = {'alpha': alphas} # put your gradient descent parameters here
print params
gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=50),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l2 in gs.grid_scores_:
    print l2

{'alpha': array([  1.00000000e-10,   1.00000000e-09,   1.00000000e-08,
         1.00000000e-07,   1.00000000e-06,   1.00000000e-05,
         1.00000000e-04,   1.00000000e-03,   1.00000000e-02,
         1.00000000e-01])}
BEST ESTIMATOR
1673.4711707
SGDRegressor(alpha=9.9999999999999995e-07, average=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='invscaling', loss='squared_loss', n_iter=50,
       penalty='l2', power_t=0.25, random_state=None, shuffle=True,
       verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -1675.57559, std: 151.01367, params: {'alpha': 1e-10}
mean: -1675.06859, std: 151.80016, params: {'alpha': 1.0000000000000001e-09}
mean: -1675.21180, std: 152.57762, params: {'alpha': 1e-08}
mean: -1674.02517, std: 151.99124, params: {'alpha': 9.9999999999999995e-08}
mean: -1673.47117, std: 152.49810, params: {'alpha': 9.9999999999999995e-07}
mean: -1674.05008, std: 152.09132, params: {'alpha': 1.0000000000000001e-05}
mean: -1674.08

Conclusion: By only altering the alpha value and running the above code many times, we find that the MSE does not change much in between 1e-10 to 1e-4. The best estimator changes every time. MSE starts to rise when alpha is set beyond 1e-4. I also change the number of iterations in order to compare with the results below obtained from Lasso.

In [56]:
alphas = np.logspace(-10, -1, 10)
l1s = np.arange(0, 1, 0.05)
params = {'alpha': alphas, 'l1_ratio': l1s} # put your gradient descent parameters here
print params
gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=5, penalty='l1'),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l1 in gs.grid_scores_:
    print l1

{'alpha': array([  1.00000000e-10,   1.00000000e-09,   1.00000000e-08,
         1.00000000e-07,   1.00000000e-06,   1.00000000e-05,
         1.00000000e-04,   1.00000000e-03,   1.00000000e-02,
         1.00000000e-01]), 'l1_ratio': array([ 0.  ,  0.05,  0.1 ,  0.15,  0.2 ,  0.25,  0.3 ,  0.35,  0.4 ,
        0.45,  0.5 ,  0.55,  0.6 ,  0.65,  0.7 ,  0.75,  0.8 ,  0.85,
        0.9 ,  0.95])}
BEST ESTIMATOR
1688.07754617
SGDRegressor(alpha=1.0000000000000001e-05, average=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.60000000000000009,
       learning_rate='invscaling', loss='squared_loss', n_iter=5,
       penalty='l1', power_t=0.25, random_state=None, shuffle=True,
       verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -1688.99274, std: 83.46992, params: {'alpha': 1e-10, 'l1_ratio': 0.0}
mean: -1689.54038, std: 84.92972, params: {'alpha': 1e-10, 'l1_ratio': 0.050000000000000003}
mean: -1688.71744, std: 84.29532, params: {'alpha': 1e-10, 'l1_ratio': 0.1000000000

In [57]:
alphas = np.logspace(-10, -1, 10)
l1s = np.arange(0, 1, 0.05)
params = {'alpha': alphas, 'l1_ratio': l1s} # put your gradient descent parameters here
print params
gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=50, penalty='l1'),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l1 in gs.grid_scores_:
    print l1

{'alpha': array([  1.00000000e-10,   1.00000000e-09,   1.00000000e-08,
         1.00000000e-07,   1.00000000e-06,   1.00000000e-05,
         1.00000000e-04,   1.00000000e-03,   1.00000000e-02,
         1.00000000e-01]), 'l1_ratio': array([ 0.  ,  0.05,  0.1 ,  0.15,  0.2 ,  0.25,  0.3 ,  0.35,  0.4 ,
        0.45,  0.5 ,  0.55,  0.6 ,  0.65,  0.7 ,  0.75,  0.8 ,  0.85,
        0.9 ,  0.95])}
BEST ESTIMATOR
1673.42952178
SGDRegressor(alpha=1.0000000000000001e-09, average=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.050000000000000003,
       learning_rate='invscaling', loss='squared_loss', n_iter=50,
       penalty='l1', power_t=0.25, random_state=None, shuffle=True,
       verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -1674.41568, std: 75.62859, params: {'alpha': 1e-10, 'l1_ratio': 0.0}
mean: -1673.91789, std: 75.60558, params: {'alpha': 1e-10, 'l1_ratio': 0.050000000000000003}
mean: -1673.61311, std: 75.47651, params: {'alpha': 1e-10, 'l1_ratio': 0.10000000

Conclusion: From what is discussed in class, people usually use Ridge (L2) when there are more observations than features, just like our modeldata, which consists of 17378 observations and 5 features only. However, from the above experiments, I find that there are not that much difference between using Lasso or Ridge. MSE is minimized in roughly the same pace as n_iter increases (here I show n_iter = 5 and 50). L1_ratio does not affect the result for each alpha I use. However, for larger alphas (e.g. 1e-1), Lasso performs better than Ridge.

In [63]:
eta0s = np.logspace(-10, 2, 13)
params = {'eta0' : eta0s}

gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=50),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l2 in gs.grid_scores_:
    print l2

BEST ESTIMATOR
1673.70850743
SGDRegressor(alpha=0.0001, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', n_iter=50, penalty='l2', power_t=0.25,
       random_state=None, shuffle=True, verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -3703.62012, std: 98.66110, params: {'eta0': 1e-10}
mean: -3703.45285, std: 98.65998, params: {'eta0': 1.0000000000000001e-09}
mean: -3701.78131, std: 98.64855, params: {'eta0': 1e-08}
mean: -3685.18065, std: 98.54057, params: {'eta0': 9.9999999999999995e-08}
mean: -3530.04602, std: 97.45148, params: {'eta0': 9.9999999999999995e-07}
mean: -2692.86906, std: 88.59556, params: {'eta0': 1.0000000000000001e-05}
mean: -2225.53173, std: 75.71625, params: {'eta0': 0.0001}
mean: -1748.67156, std: 70.89127, params: {'eta0': 0.001}
mean: -1673.70851, std: 67.48736, params: {'eta0': 0.01}
mean: -1680.49660, std: 66.04562, params: {'eta0': 0.10000000000000001}
mean: -1683.01051, std

In [68]:
# Testing on eta0

eta0s = np.logspace(-3, 0, 13)
params = {'eta0' : eta0s}

gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=50),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l2 in gs.grid_scores_:
    print l2

BEST ESTIMATOR
1674.32071031
SGDRegressor(alpha=0.0001, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', n_iter=50, penalty='l2', power_t=0.25,
       random_state=None, shuffle=True, verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -1749.27782, std: 129.18276, params: {'eta0': 0.001}
mean: -1689.17670, std: 122.64226, params: {'eta0': 0.0017782794100389228}
mean: -1675.56223, std: 118.82469, params: {'eta0': 0.0031622776601683794}
mean: -1674.65705, std: 117.34678, params: {'eta0': 0.005623413251903491}
mean: -1674.32071, std: 117.74774, params: {'eta0': 0.01}
mean: -1676.17259, std: 115.12102, params: {'eta0': 0.017782794100389229}
mean: -1676.27391, std: 120.21613, params: {'eta0': 0.031622776601683791}
mean: -1684.69838, std: 105.97641, params: {'eta0': 0.056234132519034911}
mean: -1683.17525, std: 109.31637, params: {'eta0': 0.10000000000000001}
mean: -1678.58726, std: 121.12560, params: {'eta

The learning rate, eta0, should be around 0.01 (0.003 ~ 0.03 are all acceptable values)

In [69]:
# Repeating the same using L1

eta0s = np.logspace(-3, 0, 13)
params = {'eta0' : eta0s}

gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(n_iter=50, penalty='l1'),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='neg_mean_squared_error',
    )

gs.fit(modeldata, y)

print 'BEST ESTIMATOR'
print -gs.best_score_
print gs.best_estimator_
print 'ALL ESTIMATORS'
for l1 in gs.grid_scores_:
    print l1

BEST ESTIMATOR
1673.97798121
SGDRegressor(alpha=0.0001, average=False, epsilon=0.1,
       eta0=0.005623413251903491, fit_intercept=True, l1_ratio=0.15,
       learning_rate='invscaling', loss='squared_loss', n_iter=50,
       penalty='l1', power_t=0.25, random_state=None, shuffle=True,
       verbose=0, warm_start=False)
ALL ESTIMATORS
mean: -1748.42228, std: 92.46752, params: {'eta0': 0.001}
mean: -1688.13414, std: 91.27439, params: {'eta0': 0.0017782794100389228}
mean: -1674.63451, std: 90.91249, params: {'eta0': 0.0031622776601683794}
mean: -1673.97798, std: 91.08631, params: {'eta0': 0.005623413251903491}
mean: -1674.19443, std: 90.45945, params: {'eta0': 0.01}
mean: -1674.70289, std: 92.13233, params: {'eta0': 0.017782794100389229}
mean: -1674.88824, std: 89.65987, params: {'eta0': 0.031622776601683791}
mean: -1676.97523, std: 86.04203, params: {'eta0': 0.056234132519034911}
mean: -1675.22011, std: 91.34700, params: {'eta0': 0.10000000000000001}
mean: -1686.67799, std: 96.79900, 

Results are very similar to that using L2.

Conclusion: The stochastic gradient descent approach (SGD) seems to be outperform Ridge as we can minimize MSE in each iteration. MSE falls to a local minimum around 1673 after about 5-10 iterations. Previously, Ridge produces MSE of about 1814. Using the SGD also allows me to test different approach (L2, L1, Elastic Net). Gradient descent can take a lot of time and computationally expensive when the data is huge and we do not know exactly how large it is and thus set a large number of iternations to ensure a good result.